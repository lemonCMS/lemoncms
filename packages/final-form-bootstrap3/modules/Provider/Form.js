import React from "react";
import PropTypes from "prop-types";
import { Form as FinalForm } from "react-final-form";
import arrayMutators from "final-form-arrays";
import ContextWrapper from "../Context/ContextWrapper";

const Form = props => {
  const { children, className, debug, readOnly, listen, ...formProps } = props;

  return (
    <FinalForm
      {...formProps}
      mutators={{
        ...arrayMutators
      }}
      render={({ handleSubmit, ...rest }) => {
        return (
          <ContextWrapper
            {...rest}
            debug={debug}
            listen={listen}
            readOnly={readOnly}
          >
            <form onSubmit={handleSubmit} className={className}>
              {children}
            </form>
          </ContextWrapper>
        );
      }}
    />
  );
};

Form.propTypes = {
  className: PropTypes.string,
  validate: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  debug: PropTypes.bool,
  readOnly: PropTypes.bool,
  listen: PropTypes.func,
  initialValues: PropTypes.objectOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.array,
      PropTypes.object,
      PropTypes.bool
    ])
  )
};

Form.defaultProps = {
  className: "",
  validate: () => {},
  listen: () => {},
  debug: false,
  readOnly: false,
  initialValues: {}
};

export default Form;
