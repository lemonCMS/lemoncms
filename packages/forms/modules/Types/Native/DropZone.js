import React from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import context from '../../decorators/context';
import field from '../../decorators/field';
import FileUpload from '../_shared/FileUpload';

class DropZone extends React.Component {
  state = {
    queue: []
  };

  stack = [];

  shouldComponentUpdate(nextProps, nextState) {
    const { queue } = this.state;
    return queue !== nextState.queue;
  }

  componentWillUnmount() {
    const { queue } = this.state;
    queue.forEach(file => {
      URL.createObjectURL(file);
    });
  }

  filesFromClient = files => {
    const { queue } = this.state;
    const tmpQueue = [...queue];
    files.forEach(file => {
      file.preview = URL.createObjectURL(file);
      tmpQueue.push(file);
    });
    this.setState({ queue: tmpQueue });
  };

  addOnStack = file => {
    const { input } = this.props;
    const {
      name,
      size,
      type,
      lastModified
    } = file;

    this.stack.push({
      name,
      size,
      type,
      lastModified
    });

    input.onChange(this.stack);
  };

  removeFromStack = fileToDelete => {
    const { queue } = this.state;
    const { input } = this.props;
    const filtered = queue.filter(file => (file.name !== fileToDelete.name));
    const stackFiltered = this.stack.filter(file => (file.name !== fileToDelete.name));
    this.stack = stackFiltered;
    input.onChange(stackFiltered);
    this.setState({ queue: filtered });
  };

  displayFiles = () => {
    const { queue } = this.state;
    const { endPoint } = this.props;
    if (queue.length === 0) {
      return null;
    }

    const list = queue.map(file => (
      <FileUpload
        key={file.name}
        file={file}
        addOnStack={this.addOnStack}
        removeFromStack={this.removeFromStack}
        endPoint={endPoint}
      />
    ));

    return (
      <table className="table">
        <tbody>
          {list}
        </tbody>
      </table>
    );
  };

  render() {
    return (
      <>
        <Dropzone onDrop={this.filesFromClient}>
          {({ getRootProps, getInputProps }) => (
            <section className="file-upload-dropzone">
              <div {...getRootProps()}>
                <input {...getInputProps()} />
                <button
                  type="button"
                  className="btn btn-primary"
                >
                  select files...
                </button>
              </div>
            </section>
          )}
        </Dropzone>
        {this.displayFiles()}
      </>
    );
  }
}

DropZone.propTypes = {
  input: PropTypes.oneOfType([PropTypes.object]),
  placeholder: PropTypes.string,
  endPoint: PropTypes.shape({
    path: PropTypes.string.isRequired,
    headers: PropTypes.shape({})
  }).isRequired,
  formControl: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired
};

DropZone.defaultProps = {
  input: {},
  placeholder: null,
  formControl: null,
};

export default context(field(DropZone));
