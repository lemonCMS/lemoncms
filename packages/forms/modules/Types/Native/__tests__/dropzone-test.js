import React from 'react';
import ReactDOM from 'react-dom';
import {
  act,
  Simulate
} from 'react-dom/test-utils';
import Form from '../../../Provider/Form';
import DropZone from '../DropZone';

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe('Message', () => {

  it('render', () => {
    const submit = jest.fn();
    const file = new File(['(⌐□_□)'], 'chucknorris.png', {
      type: 'text/csv',
      size: 1024
    });
    const Component = () => <Form onSubmit={submit}>
      <DropZone placeholder={'dropzone'} name={'dropzone'} endPoint={{ path: '/api/upload' }} />
      <button type={'submit'}>do it</button>
    </Form>;

    act(() => {
      ReactDOM.render(<Component />, container);
    });
    console.log(document.body.innerHTML);
    const dropzone = container.querySelector('section');
    expect(dropzone).toBeTruthy();
    const cOU = global.URL.createObjectURL = jest.fn();
    cOU.mockReturnValueOnce('none');
    const fileInput = container.querySelector('input[type=file]');
    expect(fileInput).toBeTruthy();
    // input.files is a read-only property
    // so this is not allowed
    // input.files = [file]

    // But DOM properties are reconfigurable
    // I got this while reading through a related JSDOM Github issue
    // https://github.com/jsdom/jsdom/issues/1272#issuecomment-150670691
    act(() => {
      Object.defineProperty(fileInput, 'files', {
        value: [file],
      });
    })
    act(() => {

      Simulate.change(fileInput);
    });

    act(() => {
      // dropzone.dispatchEvent(new MouseEvent('drop', { bubbles: true }));
    });

    console.log(document.body.innerHTML);

  });
});
