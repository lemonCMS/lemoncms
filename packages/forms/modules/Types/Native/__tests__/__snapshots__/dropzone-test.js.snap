// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="ffc-field"
  >
    <section
      className="file-upload-dropzone"
    >
      <div
        onBlur={[Function]}
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <input
          autoComplete="off"
          multiple={true}
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          tabIndex={-1}
          type="file"
        />
        <button
          className="btn btn-primary"
          type="button"
        >
          select files...
        </button>
      </div>
    </section>
  </div>
  <button
    type="submit"
  >
    do it
  </button>
</form>
`;
